File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java
package org.wordpress.android.e2e.pages;

import androidx.test.espresso.Espresso;
import androidx.test.espresso.ViewInteraction;
import androidx.test.espresso.matcher.ViewMatchers.Visibility;

import org.wordpress.android.R;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;
import static org.wordpress.android.support.WPSupportUtils.clickOn;
import static org.wordpress.android.support.WPSupportUtils.getCurrentActivity;
import static org.wordpress.android.support.WPSupportUtils.isElementDisplayed;
import static org.wordpress.android.support.WPSupportUtils.scrollToThenClickOn;
import static org.wordpress.android.support.WPSupportUtils.waitForElementToBeDisplayed;

public class MePage {
    // Labels
    private static ViewInteraction displayName = onView(withId(R.id.me_display_name));
    private static ViewInteraction usernameLabel = onView(withId(R.id.me_username));

    // Buttons
    private static ViewInteraction appSettings = onView(withId(R.id.row_app_settings));

    public MePage() {
    }

    public MePage go() {
        // Using the settings button as a marker for successfully navigating to the page
        while (!isElementDisplayed(appSettings)) {
            clickOn(R.id.nav_sites);
            clickOn(onView(allOf(withId(R.id.me_item), withEffectiveVisibility(Visibility.VISIBLE))));
        }

        if (!isSelfHosted()) {
            displayName.check(matches(isDisplayed()));
        }

        return this;
    }

    public void goBack() {
        Espresso.pressBack();
    }

    public MePage verifyUsername(String username) {
        String atUsername = "@" + username;
        usernameLabel.check(matches(withText(atUsername)));

        return this;
    }

    public boolean isSelfHosted() {
        waitForElementToBeDisplayed(R.id.row_logout);
        return isElementDisplayed(onView(withText(R.string.sign_in_wpcom)));
    }

    public void openAppSettings() {
        appSettings.perform(click());
    }

    public void logout() {
        ViewInteraction logOutButton = onView(allOf(
                withId(R.id.me_login_logout_text_view),
                withText(getCurrentActivity().getString(R.string.me_disconnect_from_wordpress_com))));
        waitForElementToBeDisplayed(logOutButton);
        while (!isElementDisplayed(android.R.id.button1)) {
            scrollToThenClickOn(logOutButton);
        }
        clickOn(android.R.id.button1);
    }
}

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.kt
package org.wordpress.android.e2e.pages

import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import org.hamcrest.Matchers
import org.wordpress.android.R
import org.wordpress.android.support.WPSupportUtils
import android.R as AndroidR

class MePage {
    fun go(): MePage {
        // Using the settings button as a marker for successfully navigating to the page
        while (!WPSupportUtils.isElementDisplayed(appSettings)) {
            WPSupportUtils.clickOn(R.id.nav_me)
        }
        if (!isSelfHosted) {
            displayName.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
        return this
    }

    fun goBack() {
        Espresso.pressBack()
    }

    fun verifyUsername(username: String): MePage {
        val atUsername = "@$username"
        usernameLabel.check(ViewAssertions.matches(ViewMatchers.withText(atUsername)))
        return this
    }

    val isSelfHosted: Boolean
        get() {
            WPSupportUtils.waitForElementToBeDisplayed(R.id.row_app_settings)
            return WPSupportUtils.isElementDisplayed(Espresso.onView(ViewMatchers.withText(R.string.sign_in_wpcom)))
        }

    fun openAppSettings() {
        appSettings.perform(ViewActions.click())
    }

    fun logout() {
        val logOutButton = Espresso.onView(
            Matchers.allOf(
                ViewMatchers.withId(R.id.me_login_logout_text_view),
                ViewMatchers.withText(
                    WPSupportUtils.getCurrentActivity()
                        .getString(R.string.me_disconnect_from_wordpress_com)
                )
            )
        )
        while (!WPSupportUtils.isElementDisplayed(AndroidR.id.button1)) {
            WPSupportUtils.scrollToThenClickOn(logOutButton)
        }
        WPSupportUtils.clickOn(AndroidR.id.button1)
    }

    companion object {
        // Labels
        private val displayName = Espresso.onView(ViewMatchers.withId(R.id.me_display_name))
        private val usernameLabel = Espresso.onView(ViewMatchers.withId(R.id.me_username))

        // Buttons
        private val appSettings = Espresso.onView(ViewMatchers.withId(R.id.row_app_settings))
    }
}
